Gem/Swap/ViewModel/SwapViewModel.swift

```diff
@@ -63,152 +63,174 @@ class SwapViewModel: ObservableObject {
+                transferData = try getSwapDataOnApprove(fromAsset: fromAsset, toAsset: toAsset, spender: spender, spenderName: swapAvailability.quote.provider.name)
             }
-            NSLog("allowance \(allowance)")
-        case .trc20, .ibc, .jetton, .synth:
-            fatalError()
+        } catch {
+            swapAvailabilityState = .error(error)
         }
     }
-    
-    func getAllowanceData(fromAsset: Asset, toAsset: Asset, spender: String, spenderName: String) throws -> TransferData {
-        return TransferData(
-            type: .swap(
-                fromAsset,
-                toAsset,
-                SwapAction.approval(spender: spender, allowance: .MAX_256)
-            ),
-            recipientData: RecipientData(
-                asset: fromAsset,
-                recipient: Recipient(name: spenderName, address: try fromAsset.getTokenId(), memo: .none)
-            ),
-            value: BigInt.zero
-        )
+}
+
+// MARK: - Private
+
+extension SwapViewModel {
+    private func isValidFromValue(assetData: AssetData) -> Bool {
+        guard !fromValue.isEmpty else { return false }
+        let asset = assetData.asset
+        let amount = (try? formatter.inputNumber(from: fromValue, decimals: Int(asset.decimals))) ?? BigInt.zero
+
+        return amount > 0 && amount <= assetData.balance.available
     }
-    
-    // used to fetch quote estimates. without any data
-    func quote(fromAsset: Asset, toAsset: Asset, amount: String) async throws -> SwapQuote {
-        let request = try self.request(fromAsset, toAsset, amount, includeData: false)
-        return try await service.getQuote(request: request).quote
+
+    private func getSwapDataOnApprove(fromAsset: Asset, toAsset: Asset, spender: String, spenderName: String) throws -> TransferData {
+        let transferDataType: TransferDataType = .swap(fromAsset, toAsset, SwapAction.approval(spender: spender, allowance: .MAX_256))
+        let recipientData = RecipientData(asset: fromAsset,
+                                          recipient: Recipient(name: spenderName, address: try fromAsset.getTokenId(),
+                                                               memo: .none))
+
+        return TransferData(type: transferDataType, recipientData: recipientData, value: BigInt.zero)
     }
 
-    // used to fetch quote data for swaps
-    func swap(fromAsset: Asset, toAsset: Asset, amount: String) async throws -> TransferData {
-        let request = try self.request(fromAsset, toAsset, amount, includeData: true)
-        
-        let quote = try await service.getQuote(request: request).quote
+    private func getSwapData(fromAsset: Asset, toAsset: Asset, amount: String) async throws -> TransferData {
+        let quote = try await self.quote(fromAsset: fromAsset, toAsset: toAsset, amount: amount, includeData: true)
         guard let data = quote.data else {
-            throw AnyError("")
+            throw AnyError("No Quote data")
```

Comment:
```suggestion
            throw SwapError.noQuoteData
```

let's create SwapError enum and throw SwapError.noQuoteData